colossalai/zero/sharded_optim/sharded_adam.py

```diff
@@ -0,0 +1,163 @@
+                                             growth_interval=growth_interval,
+                                             hysteresis=hysteresis,
+                                             max_scale=max_scale)
+        self._found_overflow: Tensor = torch.FloatTensor([0]).to(self.state_device)
+
+        # Early state initialization
+        for group in adam_optim.param_groups:
+            for p in group['params']:
+                state_shape = p.shape
+                if hasattr(p, 'ca_attr'):
+                    assert p.ca_attr.is_sharded, 'ShardedAdam can be only used with sharded model'
+                    # TODO: use payload shape
+                    state_shape = p.ca_attr.payload(self.state_device)
+                state = adam_optim.state[p]
+                assert len(state) == 0, 'adam optimizer initialized'
+                state['step'] = 0
+                # Exponential moving average of gradient values
+                state['exp_avg'] = torch.zeros(state_shape,
+                                               memory_format=torch.preserve_format,
+                                               dtype=torch.float,
+                                               device=self.state_device)
+                # Exponential moving average of squared gradient values
+                state['exp_avg_sq'] = torch.zeros(state_shape,
+                                                  memory_format=torch.preserve_format,
+                                                  dtype=torch.float,
+                                                  device=self.state_device)
+                if group['amsgrad']:
+                    # Maintains max of all exp. moving avg. of sq. grad. values
+                    state['max_exp_avg_sq'] = torch.zeros(state_shape,
+                                                          memory_format=torch.preserve_format,
+                                                          dtype=torch.float,
+                                                          device=self.state_device)
+
+    def step(self, *args, **kwargs):
+        # unscale grads if scaled
+        if self.optim_state == OptimState.SCALED:
+            self._unscale_grads()
+
+        found_inf = self._check_overflow()
+        self.grad_scaler.update(found_inf)
+
+        if found_inf:
+            self.zero_grad()
+            return
+
+        # Write payload back to p.data
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                data = p.data
+                if hasattr(p, 'ca_attr'):
+                    data = p.ca_attr.payload(self.state_device)
+                if torch.is_floating_point(data) and data.dtype != torch.float:
+                    data = data.to(torch.float)
+                p.data = data
+        ret = self.optim.step(*args, **kwargs)
+        # Set p.data to None
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                p.data = None
```

Comment:
@feifeibear I should write the updated weight to sharded param as well. Do we have a method to update it?